File: java-soa/order-service/src/main/java/com/ecommerce/order/application/OrderServiceImpl.java
package com.ecommerce.order.application;

import com.ecommerce.order.domain.model.Order;
import com.ecommerce.order.domain.service.OrderService;
import com.ecommerce.order.infrastructure.messaging.OrderEventPublisher;
import com.ecommerce.order.infrastructure.persistence.OrderPersistenceAdapter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

/**
 * Order Service Implementation - Application Layer
 * 
 * <p><b>Saga Pattern - Orchestrator</b>: This service acts as the saga orchestrator
 * for the order creation flow. It initiates the saga by publishing events and
 * handles responses from other services.</p>
 * 
 * <p><b>Flow</b>:</p>
 * <ol>
 *   <li>createOrder() → saves order as PENDING → publishes order.created</li>
 *   <li>confirmOrder() → updates to CONFIRMED → publishes order.confirmed</li>
 *   <li>cancelOrder() → updates to CANCELLED → publishes order.cancelled</li>
 * </ol>
 * 
 * <p><b>Strategy Pattern</b>: The order processing strategy could be swapped
 * (e.g., synchronous vs. asynchronous confirmation) by implementing a different
 * version of this service.</p>
 */
@Service
@Transactional
public class OrderServiceImpl implements OrderService {

    private static final Logger log = LoggerFactory.getLogger(OrderServiceImpl.class);

    private final OrderPersistenceAdapter persistenceAdapter;
    private final OrderEventPublisher eventPublisher;

    public OrderServiceImpl(OrderPersistenceAdapter persistenceAdapter,
                            OrderEventPublisher eventPublisher) {
        this.persistenceAdapter = persistenceAdapter;
        this.eventPublisher = eventPublisher;
    }

    /**
     * Creates a new order and initiates the Saga.
     * 
     * <p><b>Saga Step 1</b>: Order is created with PENDING status.
     * The OrderCreatedEvent is published to Kafka, which triggers
     * the Inventory Service to check and reserve stock.</p>
     */
    @Override
    public Order createOrder(Order order) {
        log.info("Creating new order: userId={}, items={}", 
                order.getUserId(), order.getItems().size());

        // Save order with PENDING status
        Order savedOrder = persistenceAdapter.save(order);

        // Publish event to ESB - initiates the saga
        eventPublisher.publishOrderCreated(savedOrder);

        log.info("Order created and saga initiated: orderId={}, status={}",
                savedOrder.getId(), savedOrder.getStatus());
        return savedOrder;
    }

    @Override
    @Transactional(readOnly = true)
    public Optional<Order> getOrderById(UUID id) {
        log.debug("Fetching order by id: {}", id);
        return persistenceAdapter.findById(id);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Order> getOrdersByUserId(UUID userId) {
        log.debug("Fetching orders for user: {}", userId);
        return persistenceAdapter.findByUserId(userId);
    }

    @Override
    @Transactional(readOnly = true)
    public List<Order> getAllOrders() {
        log.debug("Fetching all orders");
        return persistenceAdapter.findAll();
    }

    /**
     * Confirms an order after stock reservation.
     * 
     * <p><b>Saga Step 3</b>: Called when the Inventory Service confirms
     * stock reservation (stock.reserved event received).</p>
     */
    @Override
    public void confirmOrder(UUID orderId) {
        log.info("Confirming order: orderId={}", orderId);

        Order order = persistenceAdapter.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found: " + orderId));

        // Domain business rule - validates state transition
        order.confirm();

        persistenceAdapter.save(order);

        // Publish confirmation event
        eventPublisher.publishOrderConfirmed(order);

        log.info("Order confirmed: orderId={}", orderId);
    }

    /**
     * Cancels an order (compensating transaction).
     * 
     * <p><b>Saga Compensation</b>: Called when stock is insufficient
     * or when the user explicitly cancels. Publishes order.cancelled
     * so the Inventory Service can release any reserved stock.</p>
     */
    @Override
    public void cancelOrder(UUID orderId) {
        log.info("Cancelling order: orderId={}", orderId);

        Order order = persistenceAdapter.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found: " + orderId));

        // Domain business rule - validates state transition
        order.cancel();

        persistenceAdapter.save(order);

        // Publish cancellation event - triggers stock release
        eventPublisher.publishOrderCancelled(order);

        log.info("Order cancelled: orderId={}", orderId);
    }
}

File: dotnet-soa/OrderService/Application/OrderServiceImpl.cs
using OrderService.Domain.Interfaces;
using OrderService.Domain.Models;
using OrderService.Infrastructure.Messaging;
using OrderService.Infrastructure.Persistence;

namespace OrderService.Application;

/// <summary>
/// Order service implementation - Saga Orchestrator.
/// 
/// Design Pattern: Saga - orchestrates a distributed transaction across
/// multiple services (Order + Inventory) using compensating actions.
/// 
/// Saga Flow:
/// 1. CreateOrder → publishes "order.created" → Inventory reserves stock
/// 2a. Stock reserved → publishes "order.confirmed" (happy path)
/// 2b. Insufficient stock → publishes "order.cancelled" → Inventory releases stock
/// </summary>
public class OrderServiceImpl : IOrderService
{
    private readonly OrderRepository _repository;
    private readonly KafkaProducer _kafkaProducer;
    private readonly ILogger<OrderServiceImpl> _logger;

    public OrderServiceImpl(
        OrderRepository repository,
        KafkaProducer kafkaProducer,
        ILogger<OrderServiceImpl> logger)
    {
        _repository = repository;
        _kafkaProducer = kafkaProducer;
        _logger = logger;
    }

    /// <inheritdoc />
    public async Task<Order> CreateOrderAsync(string customerName, string customerEmail, List<OrderItem> items)
    {
        var totalAmount = items.Sum(i => i.Quantity * i.UnitPrice);

        var order = new Order
        {
            CustomerName = customerName,
            CustomerEmail = customerEmail,
            Status = OrderStatus.PENDING,
            TotalAmount = totalAmount,
            Items = items
        };

        var created = await _repository.CreateAsync(order);

        // Saga Step 1: Publish order.created event to trigger stock reservation
        try
        {
            await _kafkaProducer.PublishAsync("order.created", created.Id.ToString(), new OrderCreatedEvent
            {
                OrderId = created.Id,
                CustomerName = created.CustomerName,
                CustomerEmail = created.CustomerEmail,
                Items = created.Items.Select(i => new OrderItemEvent
                {
                    ProductId = i.ProductId,
                    ProductName = i.ProductName,
                    Quantity = i.Quantity,
                    UnitPrice = i.UnitPrice
                }).ToList(),
                TotalAmount = created.TotalAmount,
                Timestamp = DateTime.UtcNow
            });
            _logger.LogInformation("Saga initiated: Published order.created for order {OrderId}", created.Id);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to publish order.created event for order {OrderId}", created.Id);
        }

        return created;
    }

    /// <inheritdoc />
    public async Task<IEnumerable<Order>> GetAllOrdersAsync()
    {
        return await _repository.GetAllAsync();
    }

    /// <inheritdoc />
    public async Task<Order?> GetOrderByIdAsync(Guid id)
    {
        return await _repository.GetByIdAsync(id);
    }

    /// <inheritdoc />
    public async Task HandleStockReservedAsync(Guid orderId)
    {
        _logger.LogInformation("Saga step: Stock reserved for order {OrderId}, confirming order", orderId);

        await _repository.UpdateStatusAsync(orderId, OrderStatus.CONFIRMED);

        // Saga Step 2a: Publish order.confirmed event
        try
        {
            await _kafkaProducer.PublishAsync("order.confirmed", orderId.ToString(), new OrderConfirmedEvent
            {
                OrderId = orderId,
                Timestamp = DateTime.UtcNow
            });
            _logger.LogInformation("Saga completed: Published order.confirmed for order {OrderId}", orderId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to publish order.confirmed event for order {OrderId}", orderId);
        }
    }

    /// <inheritdoc />
    public async Task HandleInsufficientStockAsync(Guid orderId, string reason)
    {
        _logger.LogWarning("Saga compensation: Insufficient stock for order {OrderId}. Reason: {Reason}", orderId, reason);

        await _repository.UpdateStatusAsync(orderId, OrderStatus.CANCELLED);

        // Saga Step 2b (Compensation): Publish order.cancelled event
        try
        {
            await _kafkaProducer.PublishAsync("order.cancelled", orderId.ToString(), new OrderCancelledEvent
            {
                OrderId = orderId,
                Reason = reason,
                Timestamp = DateTime.UtcNow
            });
            _logger.LogInformation("Saga compensated: Published order.cancelled for order {OrderId}", orderId);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to publish order.cancelled event for order {OrderId}", orderId);
        }
    }
}
